kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/containOnly.kt

```diff
@@ -0,0 +1,89 @@
+package io.kotest.matchers.collections
+
+import io.kotest.assertions.print.print
+import io.kotest.matchers.*
+import kotlin.jvm.JvmName
+import io.kotest.assertions.AssertionsConfig.maxCollectionPrintSize as printSize
+
+/**
+ * Assert that a collection contains only the given elements.
+ */
+@JvmName("shouldContainOnly_iterable")
+infix fun <T> Iterable<T>?.shouldContainOnly(expected: Iterable<T>) =
+    this?.toList() should containOnly(expected.toList())
+
+/**
+ * Assert that an array contains only the given elements.
+ */
+@JvmName("shouldContainOnly_array")
+infix fun <T> Array<T>?.shouldContainOnly(expected: Array<T>) =
+    this?.asList() should containOnly(*expected)
+
+/**
+ * Assert that a collection contains only the given elements.
+ */
+fun <T> Iterable<T>?.shouldContainOnly(vararg expected: T) =
+    this?.toList() should containOnly(*expected)
+
+/**
+ * Assert that an array contains only the given elements.
+ */
+fun <T> Array<T>?.shouldContainOnly(vararg expected: T) =
+    this?.asList() should containOnly(*expected)
+
+/**
+ * Assert that a collection contains only the given elements.
+ */
+infix fun <T, C : Collection<T>> C?.shouldContainOnly(expected: C) = this should containOnly(expected)
+
+/**
+ * Assert that a collection contains only the given elements.
+ */
+fun <T> Collection<T>?.shouldContainOnly(vararg expected: T) = this should containOnly(*expected)
+
+/**
+ * Assert that a collection contains only the given elements
+ */
+fun <T> containOnly(vararg expected: T): Matcher<Collection<T>?> = containOnly(expected.asList())
+
+/**
+ * Assert that a collection contains only the given elements.
```

Comment:
I'd like this documentation to be more explicit on how the method will do this. We can do that with a couple of code examples if you think it will be more instructive, like:

listOf(1, 1, 1).shouldContainOnly(1)   // Assertion passes
listOf(1, 2, 1).shouldContainOnly(1) // Assertion fails

This way we have a couple of samples on why we should use this method.

This applies for all the other documentations. What do you think?

Reply:
> . Shall I add the examples to all flavours? e.g array, iterable, vararg
Yes.
Although honestly, I'd just copy paste it a couple of times and be done with it 😅

Reply:
@LeoColman I have some more queries on the assertions. What should be the expectation in these 
```
list(1, 2) shouldContainOnly setOf(1, 2) // Assertion passes ?
list(1, 2) shouldNotContainOnly listOf(3) // Assertion passes ? 
```

Reply:
That's a good question.
When things don't make much sense to be used in that way I usually care less about its documentation.

So feel free to not include these cases in the documentation. However, we should define a explicit behavior, yes. As the proposer of the feature, what do you think should be the behavior?

Reply:
For, `listOf(1, 2) shouldContainOnly setOf(1, 2)` I think assertion should pass in this case since the equality is on the items and not really on what type you are passing as long as both are iterable. And, for `listOf(1, 2) shouldNotContainOnly listOf(3)` assertion should fail since the actual collection doesn't even have `3` (not able to think of any use-case though were this will be required). 

Also, I think since no one has asked for these flavours as of now. So, we can be more restrictive in what we offer on public APIs and add these later once requirements come in. Let me your thoughts especially on being restrictive upfront and just disclosing the below 2.

```
fun <T> Iterable<T>?.shouldContainOnly(vararg expected: T)
fun <T> Array<T>?.shouldContainOnly(vararg expected: 
```

@klitoskyriacou Let us know your thoughts as well

Reply:
I would treat `shouldContainOnly` so as to be consistent with `shouldContainExactly` with regards to their signatures and overloads. The main difference between the two is when the actual collection contains duplicates: `listOf(1, 1, 2).shouldContainOnly(1, 2)` would pass, whereas `shouldContainExactly` fails. So, since `listOf(1, 2) shouldContainExactly setOf(1, 2)` is well defined and it passes, I would agree that we should also support `listOf(1, 2) shouldContainOnly setOf(1, 2)` and that it should also pass.

As for `listOf(1, 2) shouldNotContainOnly listOf(3)`, I think that should pass, because:

1. `shouldNotContainOnly` passes whenever `shouldContainOnly` fails, and vice versa;
2. `listOf(1, 2) shouldContainOnly listOf(3)` would obviously fail
3. since that fails, its opposite (`listOf(1, 2) shouldNotContainOnly listOf(3)`) should pass.

Reply:
thanks, @klitoskyriacou. 

I'm done with all the suggested changes @LeoColman, @charleskorn. Please re-review